sink/src/main/kotlin/org/neo4j/connectors/kafka/sink/strategy/pattern/PatternConstraintValidator.kt

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) "Neo4j"
+ * Neo4j Sweden AB [https://neo4j.com]
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.connectors.kafka.sink.strategy.pattern
+
+import org.neo4j.connectors.kafka.data.ConstraintData
+import org.neo4j.connectors.kafka.data.ConstraintEntityType
+import org.neo4j.connectors.kafka.data.ConstraintType
+import org.neo4j.connectors.kafka.data.NODE_KEY_CONSTRAINTS
+import org.neo4j.connectors.kafka.data.RELATIONSHIP_KEY_CONSTRAINTS
+import org.neo4j.connectors.kafka.utils.ListUtils.equalsIgnoreOrder
+
+object PatternConstraintValidator {
+  fun checkNodeWarnings(constraints: List<ConstraintData>, pattern: NodePattern): List<String> {
+    val typeConstraintMap =
+        constraints
+            .filter {
+              it.entityType == ConstraintEntityType.NODE.value &&
+                  NODE_KEY_CONSTRAINTS.contains(it.constraintType) &&
+                  pattern.labels.contains(it.labelOrType)
+            }
+            .groupBy { "${it.constraintType}-${it.labelOrType}" }
+
+    val keys = pattern.keyProperties.map { it.to }
+
+    val warningMessages = mutableListOf<String>()
+
+    for (label in pattern.labels) {
+      if (checkNodeKey(label, keys, typeConstraintMap)) {
+        continue
+      }
+
+      if (checkNodeUniqueness(label, keys, typeConstraintMap) &&
+          checkNodePropertyExistence(label, keys, typeConstraintMap)) {
+        continue
+      }
+
+      warningMessages.add(
```

Comment:
Let's say I have a pattern for 2 labels: `Foo` and `Bar` and 2 keys: `foo1` and `foo2`.
I would get a warning if I only have a key constraint on `(:Foo {foo1, foo2})` because `Bar` does not have a key constraint?
If so, that seems incorrect to me. We should get a warning only if **none** of the labels have constraints matching exactly the pattern key definition.

Reply:
I think we should get at most 1 warning per pattern. Something along the lines (wording improvement ideas most welcome):

```
None of the labels "Label1", "Label2" match the keys defined by the pattern (:ALabel:BLabel{!id, !id1, id2}).
Please fix at least one the following mismatch:
 - Label1 has 1 key(s) id1 (expected 2 keys: id1, id2)
 - Label2 has 1 key(s) id2 (expected 2 keys: id1, id2)
```

Reply:
Almost :)

**On the Neo4j reading side (source)**
If you run `MATCH (n:Foo:Bar {some: "thing"}) [..]` and only have a key constraint on `:Foo{some}`, the Cypher planner is still smart enough to read data from `Foo`'s constraint underlying index. This avoids costly label scans, which is (no pun intended) key. 

**On the Neo4j writing side (sink)**
When creating new nodes, we want to make sure that at least 1 label has the constraint in place as to preserve the data integrity.